README.md
# 2048 game in rust lang

To run, you must clone the project first.

```sh
  $ git clone git@github.com:tacnoman/2048-rust.git
```

Build with:

```sh
  $ cargo build
```

And run:

```sh
  $ cargo run
```

And let's funny. You can use `w`, `a`, `s` and `d` to command the game.

Bugs?

`Tacnoman <renatocassino@gmail.com>`

## Screenshot

![Screenshot](./images/v1.1.png)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
### Added
- Better interface for the game
- More tests
- ASCII name in game
- New screenshot for readme

### Changed
- Add a changelog
- Colors in game

### Fixed
- Bug when verify if the user loose the game.

## [1.0] - 2016-11-03
### Added
- First version of the game


src/lib.rs
mod game;

#[test]
fn slide_array() {
    let mut game: [i32; 4] = [2,2,2,2];
    game::slide_array(&mut game);

    assert!(game[0] == 4);
    assert!(game[1] == 4);
    assert!(game[2] == 0);
    assert!(game[3] == 0);
}

#[test]
fn replace_random_zero_in_2_or_4_in_game() {
    let mut game: [[i32; 4]; 4] = [
        [2,2,2,2],
        [2,2,2,2],
        [2,2,2,2],
        [0,2,2,2]
    ];

    game::add_number(&mut game);
    assert!(
        ((game[3][0] == 2) || (game[3][0] == 4))
    );
}

#[test]
fn test_if_loose_the_game() {
    let game: [[i32; 4]; 4] =
    [
        [1,2,3,4],
        [5,6,7,8],
        [9,10,11,12],
        [13,14,15,16]
    ];

    assert_eq!(game::is_looser_game(&game), true);
}


#[test]
fn test_if_not_loose_the_game() {
    let game: [[i32; 4]; 4] =
    [
        [2,2,2,2],
        [2,2,2,2],
        [2,2,2,2],
        [2,2,2,0]
    ];

    assert_eq!(game::is_looser_game(&game), false);
}

#[test]
fn test_if_not_loose_the_game_with_possible_movement() {
    let game: [[i32; 4]; 4] =
    [
        [2,2,2,2],
        [2,2,2,2],
        [2,2,2,2],
        [2,2,2,2]
    ];

    assert_eq!(game::is_looser_game(&game), false);
}

#[test]
fn test_if_not_win_the_game_with_zeros() {
    let game: [[i32; 4]; 4] =
    [
        [0,0,0,0],
        [0,0,0,0],
        [0,0,0,0],
        [0,0,0,0]
    ];

    assert_eq!(game::is_winner_game(&game), false);
}


#[test]
fn test_if_win_the_game() {
    let game: [[i32; 4]; 4] =
    [
        [0,2048,0,0],
        [0,0,0,0],
        [0,0,0,0],
        [0,0,0,0]
    ];

    assert!(game::is_winner_game(&game));
}

#[test]
fn test_rotate_board() {

    /*
      You must rotate the array in your mind. The first item is column and the second is the line.
    In draw board this game is like this:

       1 |    2 |    3 |    4
       5 |    6 |    7 |    8
       9 |   10 |   11 |   12
      13 |   14 |   15 |   16

    */
    let mut game: [[i32; 4]; 4] =
    [
        [1,5,9,13],
        [2,6,10,14],
        [3,7,11,15],
        [4,8,12,16]
    ];

    /*
      Response (in draw board):

      13 |    9 |    5 |    1
      14 |   10 |    6 |    2
      15 |   11 |    7 |    3
      16 |   12 |    8 |    4

      Array:

        [13,14,15,16],
        [9,10,11,12],
        [5,6,7,8],
        [1,2,8,4]      

     */
    game::rotate_board_game(&mut game);
    assert_eq!(game[0][0], 13);
    assert_eq!(game[0][1], 14);
    assert_eq!(game[0][2], 15);
    assert_eq!(game[0][3], 16);

    assert_eq!(game[1][0], 9);
    assert_eq!(game[1][1], 10);
    assert_eq!(game[1][2], 11);
    assert_eq!(game[1][3], 12);

    assert_eq!(game[2][0], 5);
    assert_eq!(game[2][1], 6);
    assert_eq!(game[2][2], 7);
    assert_eq!(game[2][3], 8);

    assert_eq!(game[3][0], 1);
    assert_eq!(game[3][1], 2);
    assert_eq!(game[3][2], 3);
    assert_eq!(game[3][3], 4);
}



src/main.rs
mod game;
extern crate rustbox;
use rustbox::{Color, RustBox};
use rustbox::Key;

fn move_up(game: &mut[[i32; 4]; 4]) -> bool {
    let mut success = false;
    for x in 0..4 {
        success |= game::slide_array(&mut game[x]);
    }
    return success;
}

fn move_left(mut game: &mut[[i32; 4]; 4]) -> bool {
    game::rotate_board_game(&mut game);
    let success = move_up(&mut game);
    game::rotate_board_game(&mut game);
    game::rotate_board_game(&mut game);
    game::rotate_board_game(&mut game);
    return success;
}

fn move_right(mut game: &mut[[i32; 4]; 4]) -> bool {
    game::rotate_board_game(&mut game);
    game::rotate_board_game(&mut game);
    game::rotate_board_game(&mut game);
    let success = move_up(&mut game);
    game::rotate_board_game(&mut game);
    return success;    
}

fn move_down(mut game: &mut[[i32; 4]; 4]) -> bool {
    game::rotate_board_game(&mut game);
    game::rotate_board_game(&mut game);
    let success = move_up(&mut game);
    game::rotate_board_game(&mut game);
    game::rotate_board_game(&mut game);
    return success;
}

fn print_logo(rustbox: &RustBox, version: &str) {
    let default_color = Color::Default;
    let line_version = format!("                                                                 Version: {}", version);
    rustbox.print(4,1,rustbox::RB_BOLD, default_color, default_color, "    _______  _______  _   ___   _____     _______  _______  __   __  _______ ");
    rustbox.print(4,2,rustbox::RB_BOLD, default_color, default_color, "   |       ||  _    || | |   | |  _  |   |       ||   _   ||  |_|  ||       |");
    rustbox.print(4,3,rustbox::RB_BOLD, default_color, default_color, "   |____   || | |   || |_|   | | |_| |   |    ___||  |_|  ||       ||    ___|");
    rustbox.print(4,4,rustbox::RB_BOLD, default_color, default_color, "    ____|  || | |   ||       ||   _   |  |   | __ |       ||       ||   |___ ");
    rustbox.print(4,5,rustbox::RB_BOLD, default_color, default_color, "   | ______|| |_|   ||___    ||  | |  |  |   ||  ||       ||       ||    ___|");
    rustbox.print(4,6,rustbox::RB_BOLD, default_color, default_color, "   | |_____ |       |    |   ||  |_|  |  |   |_| ||   _   || ||_|| ||   |___ ");
    rustbox.print(4,7,rustbox::RB_BOLD, default_color, default_color, "   |_______||_______|    |___||_______|  |_______||__| |__||_|   |_||_______|");
    rustbox.print(4,8,rustbox::RB_BOLD, default_color, default_color, "                                                                  By Tacnoman");
    rustbox.print(4,9,rustbox::RB_BOLD, default_color, default_color, &line_version);
}

fn print_message(rustbox: &RustBox) {
    rustbox.print(8, 30, rustbox::RB_BOLD, Color::Default, Color::Default, "You can control the board with the keys w,a,s and d.");
}

fn main() {
    let version = "1..0";
    let mut game: [[i32; 4]; 4] =
        [
            [16,64,32,0], // Line is a column
            [0,0,0,512],
            [0,256,512,128],
            [1024,1024,512,256]
        ];

    // Start game with two numbers
    game::add_number(&mut game);
    game::add_number(&mut game);

    let rustbox = match RustBox::init(Default::default()) {
        Result::Ok(v) => v,
        Result::Err(e) => panic!("{}", e),
    };

    let mut winner = false;
    let mut success = false;
    loop {
        print_logo(&rustbox, &version);
        print_message(&rustbox);

        game::print_board_game(&rustbox, &game);
        rustbox.present();
        match rustbox.poll_event(false) {
            Ok(rustbox::Event::KeyEvent(key)) => {
                match key {
                    Key::Char('q') => { break; }
                    Key::Char('w') => { success = move_up(&mut game); }
                    Key::Char('a') => { success = move_left(&mut game); }
                    Key::Char('s') => { success = move_down(&mut game); }
                    Key::Char('d') => { success = move_right(&mut game); }
                    _ => { continue; }
                }
            },
            Err(e) => panic!("{}", e),
            _ => { }
        }

        if !success {
            continue;
        }

        if game::is_winner_game(&mut game) {
            winner = true;
            break;
        }

        if game::is_looser_game(&mut game) {
            break;
        }
        game::add_number(&mut game);

    }

    game::print_board_game(&rustbox, &game);

    loop {
        if winner {
            let message = " Congrats! You win the game!! Press Q to quit. ";
            rustbox.print(8, 28, rustbox::RB_BOLD, Color::Default, Color::Green, message);
        } else {
            let message = " You loose the game! Press Q to quit. ";
            rustbox.print(8, 28, rustbox::RB_BOLD, Color::Default, Color::Red, message);
        }
        rustbox.present();
        match rustbox.poll_event(false) {
            Ok(rustbox::Event::KeyEvent(key)) => {
                match key {
                    Key::Char('q') => { break; }
                    _ => { continue; }
                }
            },
            Err(e) => panic!("{}", e),
            _ => { }
        }
    }
}


src/game.rs
extern crate rand;

use self::rand::Rng;

extern crate ansi_term;

extern crate rustbox;
use self::rustbox::{Color, RustBox, Style};

pub fn rotate_board_game(game: &mut[[i32; 4]; 4]) {
    let n = 4;
    for i in 0..(4/2) {
        for j in i..n-i-1 {
            let aux = game[i][j];
            game[i][j] = game[j][n-i-1];
            game[j][n-i-1] = game[n-i-1][n-j-1];
            game[n-i-1][n-j-1] = game[n-j-1][i];
            game[n-j-1][i] = aux;
        }
    }
}

pub fn add_number(game: &mut[[i32; 4]; 4]) {
    let mut indexes: [[usize; 2]; 16] = [
        [0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0]
    ];
    let mut count = 0;

    for y in 0..4 {
        for x in 0..4 {
            if game[x][y] == 0 {
                indexes[count][0] = x;
                indexes[count][1] = y;
                count = count + 1;
            }
        }
    }

    let secret_number = rand::thread_rng().gen_range(0, count);
    let number_to_add = if (rand::thread_rng().gen_range(0, 10) % 2) == 0 { 2 } else { 4 };

    let v1:usize = indexes[secret_number][0];
    let v2:usize = indexes[secret_number][1];
    game[v1][v2] = number_to_add;
}

fn has_number_in_game(game: &[[i32;4]; 4], number:i32) -> bool {
    for y in 0..4 {
        for x in 0..4 {
            if game[x][y] == number {
                return true;
            }
        }
    }

    return false;
}

pub fn is_looser_game(game: &[[i32;4]; 4]) -> bool {
    let has_zero = has_number_in_game(game, 0);
    if has_zero {
        return false;
    }

    for y in 0..3 {
        for x in 0..3 {
            if game[x][y] == game[x+1][y] {
                return false;
            }
            if game[x][y] == game[x][y+1] {
                return false;
            }
        }
    }
    return true;
}

pub fn is_winner_game(game: &[[i32;4]; 4]) -> bool {
    return has_number_in_game(game, 2048);
}

fn find_target(game: &[i32; 4], x: i32, stop: i32) -> i32 {
    // if the position is already on the first, don't evaluate
    if x==0 {
        return x;
    }

    let mut t:usize = (x as usize) - 1;
    while t >= 0 {
        if game[t] != 0 {
            if game[t] != game[x as usize] {
                return (t as i32) + 1;
            }
            return t as i32;
        } else {
            // we should not slide further, return this one
            if t==stop as usize {
                return t as i32;
            }
        }
        t = t-1;
    }
    return x;
}


pub fn slide_array(line: &mut[i32; 4]) -> bool {
    let size = 4;
    let mut stop = 0;
    let mut success = false;

    for x in 0..size {
        if line[x] != 0 {
            let t = find_target(&line, x as i32, stop);
            if t != x as i32 {
                if line[t as usize] != 0 {
                    stop = t+1;
                }
                line[t as usize] += line[x];
                line[x] = 0;
                success = true;
            }
        }
    }
    return success;
}

fn get_number_with_pad(number: i32) -> String {
    let formated_number: String;
    if number == 0 {
        formated_number = format!("     ");
    } else if number < 10 {
        formated_number = format!("  {}  ", number.to_string());
    } else if number < 100 {
        formated_number = format!("  {} ", number.to_string());
    } else if number < 1000 {
        formated_number = format!(" {} ", number.to_string());
    } else {
        formated_number = format!(" {}", number.to_string());
    }
    return formated_number;
}

fn get_color(value: i32) -> Color {
    if value == 2 || value == 4 || value == 16 || value == 64 || value == 2048 {
        return Color::White;
    } else if value == 8 || value == 512 {
        return Color::Red;
    } else if value == 32 {
        return Color::Black;
    } else if value == 128 || value == 256 {
        return Color::Yellow;
    }
    return Color::Magenta;
}

fn get_bg(value: i32) -> Color {
    if value == 0 {
        return Color::Black;
    } else if value == 2 {
        return Color::Cyan;
    } else if value == 4 || value == 256 {
        return Color::Magenta;
    } else if value == 8 || value == 1024 {
        return Color::White;
    } else if value == 16 {
        return Color::Green;
    } else if value == 32 || value == 512 {
        return Color::Yellow;
    } else if value == 128 || value == 2048 {
        return Color::Red;
    }
    return Color::Blue;
}

fn get_font_style(value: i32) -> Style {
    if value == 128 {
        return rustbox::RB_NORMAL;
    }
    return rustbox::RB_BOLD;
}

fn print_block(rustbox: &RustBox, x: usize, y: usize, value: i32) {
    let v = get_number_with_pad(value);
    let space = "     ";

    let position_x = (x*6) + 8;
    let position_y = 13+y*4;

    let color = get_color(value);
    let bg = get_bg(value);
    let style = get_font_style(value);
    rustbox.print(position_x, position_y-1, style, color, bg, &space);
    rustbox.print(position_x, position_y, style, color, bg, &v);
    rustbox.print(position_x, position_y+1, style, color, bg, &space);

}

pub fn print_board_game(rustbox: &RustBox, game: &[[i32; 4]; 4]) {
    for y in 0..4 {
        for x in 0..4 {
            let s = game[x][y];
            print_block(&rustbox,x,y,s);
        }
    }
}
